README.md
# supermarket_sweep

Clojure app to work out how to price a shopping basket, allowing for different pricing structures and special offers.

## Usage

Currently there is only a helper function that can print out the receipt for a shopping list.

To use go to /dev/user.clj and run the `print-receipt` function, to add items to the shopping list add to the shopping-list def.

Available options are:

```onion (add :weight and value to represent the waight in kg e.g. :weight 2.457)
beans-tin (part of a special 3 for 2 offer)
coke-can (part of a special 2 for £1 offer)
large-oranges
odd-ale-bottle
brown-ale-bottle (part of a set for £6)
cask-ale-bottle (part of a set for £6)
pale-ale-bottle (part of a set for £6)
```

Known bugs: Currently the special offer for the ale doesn't work correctly, but a fix is coming soon!

## License

Copyright © 2020 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-02-29
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-02-29
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/app/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/app/compare/0.1.0...0.1.1


test/supermarket_sweep/receipt.clj
(ns supermarket-sweep.receipt
  (:require [clojure.test :refer :all]))


test/supermarket_sweep/core_test.clj
(ns supermarket-sweep.core-test
  (:require [clojure.test :refer :all]
            [supermarket-sweep.core :as target]
            [supermarket-sweep.test-data :as test-data]))

(deftest shopping-list->shopping-list+details
  (testing "prices are added to the shopping list, contains deliberate mistake
   of missing weight for onion which returns a nil"
    (let [results (target/shopping-list->shopping-list+details
                    test-data/shopping-list-small)]
      (println results)
      (is (= [{:buy 3
               :free 2
               :item "beans-tin"
               :name "Beans"
               :price-per-unit 1.75
               :promo-code "bogof-beans"
               :promotion-name "Beans 3 for 2"}
              nil
              {:buy 2
               :item "coke-can"
               :name "Coke"
               :price-per-unit 0.7
               :promo-code "two-for-one-pound-coke"
               :promo-price 1.0
               :promotion-name "Coke 2 for £1"}
              {:buy 3
               :item "brown-ale-bottle"
               :name "Brown Ale"
               :price-per-unit 2.25
               :promo-code "cure-for-what-ales-you"
               :promo-price 6.0
               :promotion-name "3 ales for £6"}
              {:item "onion"
               :name "Onions 0.535 kg @ £0.29/kg"
               :price-per-kg 0.29
               :price-per-unit 0.16
               :weight 0.535}]
             results)))))






test/supermarket_sweep/test_data.clj
(ns supermarket-sweep.test-data
  (:require [clojure.test :refer :all]))

(def shopping-list-small [{:item "beans-tin"}
                          {:item "onion"}
                          {:item "coke-can"}
                          {:item "brown-ale-bottle"}
                          {:item "onion" :weight 0.535}])

(def shopping-list [{:item "beans-tin"}
                    {:item "beans-tin"}
                    {:item "beans-tin"}
                    {:item "onion" :weight 0.535}
                    {:item "coke-can"}
                    {:item "coke-can"}
                    {:item "brown-ale-bottle"}
                    {:item "cask-ale-bottle"}
                    {:item "pale-ale-bottle"}])

(def beans
  {:item "beans-tin"
   :name "Beans"
   :price-per-unit 1.75
   :promo-code "bogof-beans"
   :promotion-name "Beans 3 for 2"
   :buy 3
   :free 2})

(def coke-can
  {:item "coke-can"
   :name "Coke"
   :price-per-unit 0.70
   :promo-code "two-for-one-pound-coke"
   :promotion-name "Coke 2 for £1"
   :promo-price 1.00
   :buy 2})

(def onions
  {:item "onion"
   :name "Onions"
   :price-per-weight 0.29
   :weight 0.435
   :price-per-unit 0.13})

(def oranges
  {:item "oranges"
   :price-per-unit 0.73})


dev/user.clj
(ns user
  (:require [supermarket-sweep.core :refer :all]
            [supermarket-sweep.promotions :refer :all]
            [supermarket-sweep.receipt :refer :all]
            [supermarket-sweep.totals :refer :all]
            [clojure.string :as string]))

(def shopping-list [{:item "beans-tin"}
                    {:item "beans-tin"}
                    {:item "beans-tin"}
                    {:item "beans-tin"}
                    {:item "onion" :weight 0.452}
                    {:item "coke-can"}
                    {:item "coke-can"}
                    {:item "brown-ale-bottle"}
                    {:item "cask-ale-bottle"}
                    {:item "pale-ale-bottle"}])

(defn print-receipt
  []
  (let [shopping-items (shopping-list->shopping-list+details shopping-list)
        sub-total (calcuate-total (map :price-per-unit shopping-items))
        discounts (shopping-list+details->discounts shopping-items)
        total-savings (calcuate-total (map :discount discounts))]
    (conj ["Thanks for shopping at Supermarket Sweep"]
          (string/join (repeat 40 "-"))
          (map (juxt :name :price-per-unit) shopping-items)
          (string/join (repeat 40 "-"))
          ["Sub total" sub-total]
          (string/join (repeat 40 "-"))
          "Savings"
          (map (juxt :name :discount) discounts)
          (string/join (repeat 40 "-"))
          ["Total savings" total-savings]
          (string/join (repeat 40 "-"))
          ["Total to pay" (+ sub-total total-savings)]
          (string/join (repeat 40 "-"))
          "Thank you, come again!")))



src/supermarket_sweep/promotions.clj
(ns supermarket-sweep.promotions)

(defmulti items->discount
          "Given an item, look up the promotion code and apply discount"
          (fn [_ items]
            (first items)))

; TODO: refactor free name to for
(defn bogof
  "calculate the bogof discount for a group of items"
  [buy free quantity price-per-unit]
  (-> (int (/ quantity buy))
      (* free)
      (+ (mod quantity buy))
      (- quantity)
      (* price-per-unit)))

(defn x-for-y-pounds
  "calculate the discount when buying multiple items for a set price
  this is a terrible way to calculate it, fix later"
  [buy promo-price quantity price-per-unit]
  (let [cost-of-packs
        (-> (int (/ quantity buy))
            (+ (mod quantity buy))
            (- quantity))
        extras (* (mod quantity buy) price-per-unit)
        total-cost-with (- (- cost-of-packs extras))
        total-cost-without (* quantity price-per-unit)]
    (Double/parseDouble
      (format "%.2f" (- (- total-cost-without total-cost-with))))))

(defmethod items->discount "bogof-beans"
  [acc items]
  (let [first-item (first (second items))
        discount (bogof (:buy first-item)
                        (:free first-item)
                        (count (second items))
                        (:price-per-unit first-item))]
    (if (< discount 0)
      (conj acc
            {:name (:promotion-name first-item)
             :discount discount}))))

(defmethod items->discount "two-for-one-pound-coke"
  [acc items]
  (let [first-item (first (second items))
        discount (x-for-y-pounds (:buy first-item)
                                 (:promo-price first-item)
                                 (count (second items))
                                 (:price-per-unit first-item))]
    (if (< discount 0)
      (conj acc
            {:name (:promotion-name first-item)
             :discount discount}))))

(defmethod items->discount "cure-for-what-ales-you"
  [acc items]
  ; do for all items as each have different prices
  (let [first-item (first (second items))
        discount (x-for-y-pounds (:buy first-item)
                                 (:promo-price first-item)
                                 (count (second items))
                                 (:price-per-unit first-item))]
    (println first-item)
    (print "discount:" discount)
    (if (< discount 0)
      (conj acc
            {:name (:promotion-name first-item)
             :discount discount}))))

(defmethod items->discount :default
  [acc _]
  acc)

(defn shopping-list+details->discounts
  [shopping-list]
  (->> shopping-list
       (group-by :promo-code)
       (reduce items->discount ())))

src/supermarket_sweep/core.clj
(ns supermarket-sweep.core)

; List of promotions
(def bogof-beans
  {:promo-code "bogof-beans"
   :promotion-name "Beans 3 for 2"
   :buy 3
   :free 2})

(def two-for-one-pound-coke
  {:promo-code "two-for-one-pound-coke"
   :promotion-name "Coke 2 for £1"
   :promo-price 1.00
   :buy 2})

(def cure-for-what-ales-you
  {:promo-code "cure-for-what-ales-you"
   :promotion-name "3 ales for £6"
   :promo-price 6.00
   :buy 3})

(defn format-double
  [double]
  (Double/parseDouble
    (format "%.2f" double)))

(defmulti item->item+details
          "Given an item, look up the price and return a map that includes the price
          and any promotions.
          In production this would be a database query but for the purpose of this
          exercise the values and promotions will be hardcoded"
          :item)

(defmethod item->item+details "beans-tin"
  [{:keys [item]}]
  (conj {:item item
         :name "Beans"
         :price-per-unit 1.75}
        bogof-beans))

; At checkout the weight is input in kg when the item is scanned
(defmethod item->item+details "onion"
  [{:keys [item weight]}]
  (if (nil? weight)
    ; raise exception here if weight hasn't been entered
    (println "Please enter weight")
    {:item item
     :name (str "Onions " weight
                " kg @ £0.29/kg")
     :price-per-kg 0.29
     :weight weight
     :price-per-unit (format-double (* 0.29 weight))}))

(defmethod item->item+details "coke-can"
  [{:keys [item]}]
  (conj {:item item
         :name "Coke"
         :price-per-unit 0.70}
        two-for-one-pound-coke))

(defmethod item->item+details "large-oranges"
  [{:keys [item]}]
  {:item item
   :name "Oranges"
   :price-per-unit 0.73})

(defmethod item->item+details "odd-ale-bottle"
  [{:keys [item]}]
  {:item item
   :name "Odd Ale"
   :price-per-unit 5.25})

(defmethod item->item+details "brown-ale-bottle"
  [{:keys [item]}]
  (conj {:item item
         :name "Brown Ale"
         :price-per-unit 2.25}
        cure-for-what-ales-you))

(defmethod item->item+details "cask-ale-bottle"
  [{:keys [item]}]
  (conj {:item item
         :name "Cask Ale"
         :price-per-unit 2.57}
        cure-for-what-ales-you))

(defmethod item->item+details "pale-ale-bottle"
  [{:keys [item]}]
  (conj {:item item
         :name "Pale Ale"
         :price-per-unit 2.99}
        cure-for-what-ales-you))

(defmethod item->item+details :default
  [{:keys [item]}]
  (str "Unknown item: " item))

(defn shopping-list->shopping-list+details
  "Given a vector of items look up item details and return a map"
  [shopping-list]
  (->> shopping-list
       (map #(item->item+details %))))
